File: impl/go/timestamp.go
package jed

import (
	"fmt"
	"math"
	"strings"
)

// Timestamp / timestamptz calendar math, parsing, and rendering (spec/design/timestamp.md).
// Both types are an i64 count of microseconds since the Unix epoch (1970-01-01 00:00:00
// UTC), proleptic Gregorian, no leap seconds. This is a §8 determinism hotspot: the
// civil↔instant conversion (Hinnant), parse grammar, and render format must be byte-identical
// across the Rust/Go/TS cores. The civil↔days path uses Go's TRUNCATING `/` paired with the
// Hinnant -399/-146096 adjustment; the instant↔civil decomposition uses FLOOR div/mod helpers.

// NegInfinity is the -infinity sentinel — the smallest i64, sorts before every finite instant.
const negInfinity int64 = -9223372036854775808

// PosInfinity is the +infinity sentinel — the largest i64, sorts after every finite instant.
const posInfinity int64 = 9223372036854775807

const (
	microsPerSec = 1_000_000
	secsPerDay   = 86_400
)

// floorDiv is integer division rounding toward negative infinity (Go's `/` truncates).
func floorDiv(a, b int64) int64 {
	q := a / b
	if (a%b != 0) && ((a < 0) != (b < 0)) {
		q--
	}
	return q
}

// floorMod is the modulo matching floorDiv (always in [0, b) for b > 0).
func floorMod(a, b int64) int64 {
	m := a % b
	if m != 0 && ((m < 0) != (b < 0)) {
		m += b
	}
	return m
}

func isLeap(y int64) bool {
	return y%4 == 0 && (y%100 != 0 || y%400 == 0)
}

func daysInMonth(y int64, month int64) int64 {
	switch month {
	case 1, 3, 5, 7, 8, 10, 12:
		return 31
	case 4, 6, 9, 11:
		return 30
	case 2:
		if isLeap(y) {
			return 29
		}
		return 28
	default:
		return 0
	}
}

// daysFromCivil returns days since 1970-01-01 for the civil date (y, m, d) (Hinnant). `y` is
// the astronomical year; `/` is truncating, paired with the y-399 adjustment (= floor for y<0).
func daysFromCivil(y, m, d int64) int64 {
	if m <= 2 {
		y--
	}
	var eraNum int64
	if y >= 0 {
		eraNum = y
	} else {
		eraNum = y - 399
	}
	era := eraNum / 400
	yoe := y - era*400
	var mAdj int64 = 9
	if m > 2 {
		mAdj = -3
	}
	doy := (153*(m+mAdj)+2)/5 + (d - 1)
	doe := yoe*365 + yoe/4 - yoe/100 + doy
	return era*146097 + doe - 719468
}

// civilFromDays returns the civil date (year, month, day) from days since 1970-01-01.
func civilFromDays(z int64) (int64, int64, int64) {
	z += 719468
	var eraNum int64
	if z >= 0 {
		eraNum = z
	} else {
		eraNum = z - 146096
	}
	era := eraNum / 146097
	doe := z - era*146097
	yoe := (doe - doe/1460 + doe/36524 - doe/146096) / 365
	y := yoe + era*400
	doy := doe - (365*yoe + yoe/4 - yoe/100)
	mp := (5*doy + 2) / 153
	d := doy - (153*mp+2)/5 + 1
	var m int64
	if mp < 10 {
		m = mp + 3
	} else {
		m = mp - 9
	}
	if m <= 2 {
		y++
	}
	return y, m, d
}

// civilFromMicros decomposes an instant into civil fields using FLOOR division (so pre-1970 /
// BC instants decompose correctly; us is always 0..999_999).
func civilFromMicros(t int64) (y, mo, d, h, mi, s, us int64) {
	us = floorMod(t, microsPerSec)
	secs := floorDiv(t, microsPerSec)
	sod := floorMod(secs, secsPerDay)
	days := floorDiv(secs, secsPerDay)
	y, mo, d = civilFromDays(days)
	h = sod / 3600
	mi = (sod % 3600) / 60
	s = sod % 60
	return
}

// --- parsing -----------------------------------------------------------------

func invalidDatetime(detail string) error {
	return newError(InvalidDatetimeFormat, detail)
}

func datetimeFieldOverflow(detail string) error {
	return newError(DatetimeFieldOverflow, detail)
}

func isWS(b byte) bool {
	return b == ' ' || b == '\t' || b == '\n' || b == '\f' || b == '\r'
}

func trimASCIIWS(s string) string {
	start, end := 0, len(s)
	for start < end && isWS(s[start]) {
		start++
	}
	for end > start && isWS(s[end-1]) {
		end--
	}
	return s[start:end]
}

// readUint reads one run of ASCII digits at *i as an i64 (checked). Empty run → 22007; a
// value that overflows i64 → 22008.
func readUint(b string, i *int) (int64, error) {
	start := *i
	var v int64
	for *i < len(b) && b[*i] >= '0' && b[*i] <= '9' {
		d := int64(b[*i] - '0')
		nv := v*10 + d
		if nv < v { // i64 overflow
			return 0, datetimeFieldOverflow("numeric field too large")
		}
		v = nv
		*i++
	}
	if *i == start {
		return 0, invalidDatetime("expected a number")
	}
	return v, nil
}

func expectByte(b string, i *int, c byte) error {
	if *i < len(b) && b[*i] == c {
		*i++
		return nil
	}
	return invalidDatetime(fmt.Sprintf("expected %q", c))
}

// readFrac parses fractional-seconds digits into microseconds (0..1_000_000; 1_000_000 means
// the rounding carried). 0–6 digits exact; 7+ round to µs half away from zero (7th digit >= 5).
func readFrac(b string, i *int) (int64, error) {
	start := *i
	for *i < len(b) && b[*i] >= '0' && b[*i] <= '9' {
		*i++
	}
	digits := b[start:*i]
	if len(digits) == 0 {
		return 0, invalidDatetime("expected fractional digits after '.'")
	}
	var us int64
	for k := 0; k < 6; k++ {
		us *= 10
		if k < len(digits) {
			us += int64(digits[k] - '0')
		}
	}
	if len(digits) > 6 && digits[6] >= '5' {
		us++
	}
	return us, nil
}

// parseDatetime parses a timestamp/timestamptz literal to µs since the epoch. For timestamptz
// (applyOffset) a trailing offset normalizes the wall clock to UTC; for timestamp an offset is
// parsed/validated but ignored (PG behavior). typeName is used only for error messages.
func parseDatetime(input string, applyOffset bool, typeName string) (int64, error) {
	s := trimASCIIWS(input)
	low := strings.ToLower(s)

	switch low {
	case "infinity", "+infinity":
		return posInfinity, nil
	case "-infinity":
		return negInfinity, nil
	}

	bc := false
	body := s
	if strings.HasSuffix(low, " bc") {
		bc = true
		body = trimASCIIWS(s[:len(s)-3])
	} else if strings.HasSuffix(low, " ad") {
		body = trimASCIIWS(s[:len(s)-3])
	}

	i := 0
	year, err := readUint(body, &i)
	if err != nil {
		return 0, err
	}
	if err := expectByte(body, &i, '-'); err != nil {
		return 0, err
	}
	month, err := readUint(body, &i)
	if err != nil {
		return 0, err
	}
	if err := expectByte(body, &i, '-'); err != nil {
		return 0, err
	}
	day, err := readUint(body, &i)
	if err != nil {
		return 0, err
	}

	bad := invalidDatetime("invalid input syntax for type " + typeName)

	var hour, minute, second, micro int64
	if i < len(body) && (body[i] == ' ' || body[i] == 'T' || body[i] == 't') {
		i++
		if hour, err = readUint(body, &i); err != nil {
			return 0, err
		}
		if err = expectByte(body, &i, ':'); err != nil {
			return 0, err
		}
		if minute, err = readUint(body, &i); err != nil {
			return 0, err
		}
		if i < len(body) && body[i] == ':' {
			i++
			if second, err = readUint(body, &i); err != nil {
				return 0, err
			}
			if i < len(body) && body[i] == '.' {
				i++
				if micro, err = readFrac(body, &i); err != nil {
					return 0, err
				}
			}
		}
	}

	var offsetSecs int64
	if i < len(body) {
		switch body[i] {
		case 'Z', 'z':
			i++
		case '+', '-':
			sign := int64(1)
			if body[i] == '-' {
				sign = -1
			}
			i++
			oh, err := readUint(body, &i)
			if err != nil {
				return 0, err
			}
			var om, os int64
			if i < len(body) && body[i] == ':' {
				i++
				if om, err = readUint(body, &i); err != nil {
					return 0, err
				}
				if i < len(body) && body[i] == ':' {
					i++
					if os, err = readUint(body, &i); err != nil {
						return 0, err
					}
				}
			}
			if oh > 15 || om > 59 || os > 59 {
				return 0, datetimeFieldOverflow("time zone offset out of range")
			}
			offsetSecs = sign * (oh*3600 + om*60 + os)
		default:
			return 0, bad
		}
	}
	if i != len(body) {
		return 0, bad
	}

	if year < 1 || year > 999_999 {
		return 0, datetimeFieldOverflow("year out of range")
	}
	if month < 1 || month > 12 {
		return 0, datetimeFieldOverflow("month out of range")
	}
	astro := year
	if bc {
		astro = 1 - year
	}
	if day < 1 || day > daysInMonth(astro, month) {
		return 0, datetimeFieldOverflow("day out of range for month")
	}
	extraDay := hour == 24 && minute == 0 && second == 0 && micro == 0
	if hour > 23 && !extraDay {
		return 0, datetimeFieldOverflow("hour out of range")
	}
	if minute > 59 {
		return 0, datetimeFieldOverflow("minute out of range")
	}
	if second > 59 {
		return 0, datetimeFieldOverflow("second out of range")
	}
	hourPart := hour
	if extraDay {
		hourPart = 0
	}

	days := daysFromCivil(astro, month, day)
	if extraDay {
		days++
	}
	micros, ok := mulAdd(days, secsPerDay, hourPart*3600+minute*60+second)
	if !ok {
		return 0, datetimeFieldOverflow("value out of range")
	}
	micros, ok = mulAdd(micros, microsPerSec, micro)
	if !ok {
		return 0, datetimeFieldOverflow("value out of range")
	}
	if applyOffset {
		off, ok2 := mul64(offsetSecs, microsPerSec)
		if !ok2 {
			return 0, datetimeFieldOverflow("value out of range")
		}
		micros, ok = sub64(micros, off)
		if !ok {
			return 0, datetimeFieldOverflow("value out of range")
		}
	}
	if micros == negInfinity || micros == posInfinity {
		return 0, datetimeFieldOverflow("value out of range")
	}
	return micros, nil
}

// mulAdd computes a*b + c with i64 overflow detection.
func mulAdd(a, b, c int64) (int64, bool) {
	p, ok := mul64(a, b)
	if !ok {
		return 0, false
	}
	return add64(p, c)
}

func mul64(a, b int64) (int64, bool) {
	if a == 0 || b == 0 {
		return 0, true
	}
	p := a * b
	if p/b != a {
		return 0, false
	}
	return p, true
}

func add64(a, b int64) (int64, bool) {
	s := a + b
	if (b > 0 && s < a) || (b < 0 && s > a) {
		return 0, false
	}
	return s, true
}

func sub64(a, b int64) (int64, bool) {
	s := a - b
	if (b < 0 && s < a) || (b > 0 && s > a) {
		return 0, false
	}
	return s, true
}

// ParseTimestamp parses a timestamp (zoneless) literal: an offset in the text is accepted and
// ignored (PG behavior).
func parseTimestamp(s string) (int64, error) { return parseDatetime(s, false, "timestamp") }

// ParseTimestamptz parses a timestamptz literal: a trailing offset normalizes the value to UTC.
func parseTimestamptz(s string) (int64, error) { return parseDatetime(s, true, "timestamptz") }

// MakeTimestamp builds a zoneless timestamp (µs since the 1970 epoch) from calendar fields — the
// workhorse for make_timestamp / make_timestamptz (functions.md §11; PG make_timestamp_internal). A
// negative year denotes BC (PG). Field validation mirrors the timestamp parser and traps 22008: the
// year magnitude in 1..999999 (no year zero), month 1..12, day valid for the month, and the
// assembled time of day not past 24:00:00 — PG allows hour = 24 or sec = 60 so long as the whole
// time of day stays within a day, enforced here by one total-of-day check. sec (double precision)
// folds to micros by one correctly-rounded multiply + half-away round (the engine's one mode —
// float.md §6); it differs from PG's rint only at an exact half-microsecond tie, which realistic
// input never hits.
func makeTimestamp(year, month, day, hour, minute int64, sec float64) (int64, error) {
	// Date fields (22008). A negative year is BC; year 0 has no AD/BC representation.
	if year == 0 {
		return 0, datetimeFieldOverflow("date field value out of range")
	}
	bc := year < 0
	mag := year // |year|; the displayed magnitude
	if bc {
		mag = -year
	}
	if mag > 999_999 {
		return 0, datetimeFieldOverflow("date field value out of range")
	}
	if month < 1 || month > 12 {
		return 0, datetimeFieldOverflow("date field value out of range")
	}
	astro := mag
	if bc {
		astro = 1 - mag
	}
	if day < 1 || day > daysInMonth(astro, month) {
		return 0, datetimeFieldOverflow("date field value out of range")
	}
	// Time fields (22008), matching PG float_time_overflows: hour 0..24, minute 0..59, sec rounded
	// to micros in [0, 60_000_000] (so sec = 60 is allowed), and the whole time of day ≤ 24:00:00.
	if hour < 0 || hour > 24 || minute < 0 || minute > 59 {
		return 0, datetimeFieldOverflow("time field value out of range")
	}
	secMicros := math.Round(sec * float64(microsPerSec)) // round-half-away (math.Round)
	if math.IsNaN(secMicros) || math.IsInf(secMicros, 0) || secMicros < 0 || secMicros > 60_000_000 {
		return 0, datetimeFieldOverflow("time field value out of range")
	}
	todMicros := ((hour*60+minute)*60)*microsPerSec + int64(secMicros) // ≤ 9e10, no i64 overflow
	if todMicros > secsPerDay*microsPerSec {
		return 0, datetimeFieldOverflow("time field value out of range")
	}
	// Compose the instant in checked i64 arithmetic; any overflow is a range error.
	micros, ok := mulAdd(daysFromCivil(astro, month, day), secsPerDay*microsPerSec, todMicros)
	if !ok {
		return 0, datetimeFieldOverflow("timestamp out of range")
	}
	if micros == negInfinity || micros == posInfinity {
		return 0, datetimeFieldOverflow("timestamp out of range") // reserved for ±infinity
	}
	return micros, nil
}

// --- rendering ---------------------------------------------------------------

func renderDatetime(micros int64, isTz bool) string {
	if micros == negInfinity {
		return "-infinity"
	}
	if micros == posInfinity {
		return "infinity"
	}
	y, mo, d, h, mi, s, us := civilFromMicros(micros)
	displayed := y
	era := ""
	if y <= 0 {
		displayed = 1 - y
		era = " BC"
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%04d-%02d-%02d %02d:%02d:%02d", displayed, mo, d, h, mi, s)
	if us != 0 {
		frac := fmt.Sprintf("%06d", us)
		frac = strings.TrimRight(frac, "0")
		b.WriteByte('.')
		b.WriteString(frac)
	}
	if isTz {
		b.WriteString("+00")
	}
	b.WriteString(era)
	return b.String()
}

// RenderTimestamp renders a timestamp value to its canonical text.
func renderTimestamp(micros int64) string { return renderDatetime(micros, false) }

// RenderTimestamptz renders a timestamptz value to its canonical text (always UTC, fixed +00).
func renderTimestamptz(micros int64) string { return renderDatetime(micros, true) }

File: impl/ts/src/timestamp.ts
// The timestamp / timestamptz calendar math, parsing, and rendering
// (spec/design/timestamp.md). Both types are an i64 count of microseconds since the Unix
// epoch (1970-01-01 00:00:00 UTC), proleptic Gregorian, no leap seconds.
//
// This is a §8 determinism hotspot: the civil↔instant conversion (Hinnant), the parse
// grammar, and the render format must be byte-identical across the Rust/Go/TS cores. ALL µs
// math is `bigint` — JS `number` loses precision past 2^53. The civil↔days path uses bigint
// `/` (which truncates toward zero, like Rust/Go) paired with the Hinnant -399/-146096
// adjustment; the instant↔civil decomposition uses the floorDiv/floorMod helpers below.

import { engineError, type EngineError } from "./errors.ts";

// NEG_INFINITY is the -infinity sentinel (the smallest i64; sorts before every finite instant).
export const NEG_INFINITY = -9223372036854775808n;
// POS_INFINITY is the +infinity sentinel (the largest i64; sorts after every finite instant).
export const POS_INFINITY = 9223372036854775807n;

const MICROS_PER_SEC = 1_000_000n;
const SECS_PER_DAY = 86_400n;

// floorDiv is bigint division rounding toward negative infinity (bigint `/` truncates).
function floorDiv(a: bigint, b: bigint): bigint {
  let q = a / b;
  if (a % b !== 0n && a < 0n !== b < 0n) q -= 1n;
  return q;
}

// floorMod is the modulo matching floorDiv (always in [0, b) for b > 0).
function floorMod(a: bigint, b: bigint): bigint {
  let m = a % b;
  if (m !== 0n && m < 0n !== b < 0n) m += b;
  return m;
}

function isLeap(y: bigint): boolean {
  return y % 4n === 0n && (y % 100n !== 0n || y % 400n === 0n);
}

export function daysInMonth(y: bigint, month: bigint): bigint {
  switch (month) {
    case 1n:
    case 3n:
    case 5n:
    case 7n:
    case 8n:
    case 10n:
    case 12n:
      return 31n;
    case 4n:
    case 6n:
    case 9n:
    case 11n:
      return 30n;
    case 2n:
      return isLeap(y) ? 29n : 28n;
    default:
      return 0n;
  }
}

// daysFromCivil returns days since 1970-01-01 for the civil date (y, m, d) (Hinnant). `y` is
// the astronomical year; `/` truncates, paired with the y-399 adjustment (= floor for y<0).
export function daysFromCivil(y: bigint, m: bigint, d: bigint): bigint {
  if (m <= 2n) y -= 1n;
  const era = (y >= 0n ? y : y - 399n) / 400n;
  const yoe = y - era * 400n;
  const doy = (153n * (m + (m > 2n ? -3n : 9n)) + 2n) / 5n + (d - 1n);
  const doe = yoe * 365n + yoe / 4n - yoe / 100n + doy;
  return era * 146097n + doe - 719468n;
}

// civilFromDays returns the civil date [year, month, day] from days since 1970-01-01.
export function civilFromDays(z: bigint): [bigint, bigint, bigint] {
  z += 719468n;
  const era = (z >= 0n ? z : z - 146096n) / 146097n;
  const doe = z - era * 146097n;
  const yoe = (doe - doe / 1460n + doe / 36524n - doe / 146096n) / 365n;
  const y = yoe + era * 400n;
  const doy = doe - (365n * yoe + yoe / 4n - yoe / 100n);
  const mp = (5n * doy + 2n) / 153n;
  const d = doy - (153n * mp + 2n) / 5n + 1n;
  const m = mp + (mp < 10n ? 3n : -9n);
  return [y + (m <= 2n ? 1n : 0n), m, d];
}

// civilFromMicros decomposes an instant into civil fields using FLOOR division (so pre-1970 /
// BC instants decompose correctly; us is always 0..999_999).
export function civilFromMicros(
  t: bigint,
): [bigint, bigint, bigint, bigint, bigint, bigint, bigint] {
  const us = floorMod(t, MICROS_PER_SEC);
  const secs = floorDiv(t, MICROS_PER_SEC);
  const sod = floorMod(secs, SECS_PER_DAY);
  const days = floorDiv(secs, SECS_PER_DAY);
  const [y, mo, d] = civilFromDays(days);
  return [y, mo, d, sod / 3600n, (sod % 3600n) / 60n, sod % 60n, us];
}

// --- parsing -----------------------------------------------------------------

export function invalidDatetime(detail: string): EngineError {
  return engineError("invalid_datetime_format", detail);
}

export function fieldOverflow(detail: string): EngineError {
  return engineError("datetime_field_overflow", detail);
}

function isWS(c: number): boolean {
  return c === 0x20 || c === 0x09 || c === 0x0a || c === 0x0c || c === 0x0d;
}

export function trimASCIIWS(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && isWS(s.charCodeAt(start))) start++;
  while (end > start && isWS(s.charCodeAt(end - 1))) end--;
  return s.slice(start, end);
}

const MAX_I64 = 9223372036854775807n;

// A small mutable cursor (index into the string).
export type Cur = { i: number };

function isDigit(c: number): boolean {
  return c >= 48 && c <= 57;
}

// readUint reads one run of ASCII digits at cur.i as a bigint. Empty run → 22007; a value
// beyond i64 → 22008.
export function readUint(b: string, cur: Cur): bigint {
  const start = cur.i;
  let v = 0n;
  while (cur.i < b.length && isDigit(b.charCodeAt(cur.i))) {
    v = v * 10n + BigInt(b.charCodeAt(cur.i) - 48);
    if (v > MAX_I64) throw fieldOverflow("numeric field too large");
    cur.i++;
  }
  if (cur.i === start) throw invalidDatetime("expected a number");
  return v;
}

export function expectChar(b: string, cur: Cur, c: string): void {
  if (cur.i < b.length && b[cur.i] === c) {
    cur.i++;
    return;
  }
  throw invalidDatetime(`expected '${c}'`);
}

// readFrac parses fractional-seconds digits into microseconds (0..1_000_000; 1_000_000 means
// the rounding carried). 0–6 digits exact; 7+ round to µs half away from zero (7th digit >= 5).
export function readFrac(b: string, cur: Cur): bigint {
  const start = cur.i;
  while (cur.i < b.length && isDigit(b.charCodeAt(cur.i))) cur.i++;
  const digits = b.slice(start, cur.i);
  if (digits.length === 0) throw invalidDatetime("expected fractional digits after '.'");
  let us = 0n;
  for (let k = 0; k < 6; k++) {
    us *= 10n;
    if (k < digits.length) us += BigInt(digits.charCodeAt(k) - 48);
  }
  if (digits.length > 6 && digits.charCodeAt(6) >= 53 /* '5' */) us += 1n;
  return us;
}

function parseDatetime(input: string, applyOffset: boolean, typeName: string): bigint {
  const s = trimASCIIWS(input);
  const low = s.toLowerCase();

  if (low === "infinity" || low === "+infinity") return POS_INFINITY;
  if (low === "-infinity") return NEG_INFINITY;

  let bc = false;
  let body = s;
  if (low.endsWith(" bc")) {
    bc = true;
    body = trimASCIIWS(s.slice(0, s.length - 3));
  } else if (low.endsWith(" ad")) {
    body = trimASCIIWS(s.slice(0, s.length - 3));
  }

  const cur: Cur = { i: 0 };
  const year = readUint(body, cur);
  expectChar(body, cur, "-");
  const month = readUint(body, cur);
  expectChar(body, cur, "-");
  const day = readUint(body, cur);

  const bad = (): EngineError => invalidDatetime(`invalid input syntax for type ${typeName}`);

  let hour = 0n;
  let minute = 0n;
  let second = 0n;
  let micro = 0n;
  if (cur.i < body.length && (body[cur.i] === " " || body[cur.i] === "T" || body[cur.i] === "t")) {
    cur.i++;
    hour = readUint(body, cur);
    expectChar(body, cur, ":");
    minute = readUint(body, cur);
    if (cur.i < body.length && body[cur.i] === ":") {
      cur.i++;
      second = readUint(body, cur);
      if (cur.i < body.length && body[cur.i] === ".") {
        cur.i++;
        micro = readFrac(body, cur);
      }
    }
  }

  let offsetSecs = 0n;
  if (cur.i < body.length) {
    const ch = body[cur.i];
    if (ch === "Z" || ch === "z") {
      cur.i++;
    } else if (ch === "+" || ch === "-") {
      const sign = ch === "-" ? -1n : 1n;
      cur.i++;
      const oh = readUint(body, cur);
      let om = 0n;
      let os = 0n;
      if (cur.i < body.length && body[cur.i] === ":") {
        cur.i++;
        om = readUint(body, cur);
        if (cur.i < body.length && body[cur.i] === ":") {
          cur.i++;
          os = readUint(body, cur);
        }
      }
      if (oh > 15n || om > 59n || os > 59n) throw fieldOverflow("time zone offset out of range");
      offsetSecs = sign * (oh * 3600n + om * 60n + os);
    } else {
      throw bad();
    }
  }
  if (cur.i !== body.length) throw bad();

  if (year < 1n || year > 999_999n) throw fieldOverflow("year out of range");
  if (month < 1n || month > 12n) throw fieldOverflow("month out of range");
  const astro = bc ? 1n - year : year;
  if (day < 1n || day > daysInMonth(astro, month))
    throw fieldOverflow("day out of range for month");
  const extraDay = hour === 24n && minute === 0n && second === 0n && micro === 0n;
  if (hour > 23n && !extraDay) throw fieldOverflow("hour out of range");
  if (minute > 59n) throw fieldOverflow("minute out of range");
  if (second > 59n) throw fieldOverflow("second out of range");
  const hourPart = extraDay ? 0n : hour;

  let days = daysFromCivil(astro, month, day);
  if (extraDay) days += 1n;
  // bigint is arbitrary-precision; range-check against i64 explicitly after composing.
  let micros = days * SECS_PER_DAY + (hourPart * 3600n + minute * 60n + second);
  micros = micros * MICROS_PER_SEC + micro;
  if (applyOffset) micros -= offsetSecs * MICROS_PER_SEC;
  if (micros < NEG_INFINITY || micros > POS_INFINITY) throw fieldOverflow("value out of range");
  if (micros === NEG_INFINITY || micros === POS_INFINITY) throw fieldOverflow("value out of range");
  return micros;
}

// parseTimestamp parses a timestamp (zoneless) literal: an offset in the text is accepted and
// ignored (PG behavior).
export function parseTimestamp(s: string): bigint {
  return parseDatetime(s, false, "timestamp");
}

// parseTimestamptz parses a timestamptz literal: a trailing offset normalizes the value to UTC.
export function parseTimestamptz(s: string): bigint {
  return parseDatetime(s, true, "timestamptz");
}

// --- rendering ---------------------------------------------------------------

function pad(n: bigint, width: number): string {
  return n.toString().padStart(width, "0");
}

function renderDatetime(micros: bigint, isTz: boolean): string {
  if (micros === NEG_INFINITY) return "-infinity";
  if (micros === POS_INFINITY) return "infinity";
  const [y, mo, d, h, mi, s, us] = civilFromMicros(micros);
  const displayed = y <= 0n ? 1n - y : y;
  const era = y <= 0n ? " BC" : "";
  let out = `${pad(displayed, 4)}-${pad(mo, 2)}-${pad(d, 2)} ${pad(h, 2)}:${pad(mi, 2)}:${pad(s, 2)}`;
  if (us !== 0n) {
    const frac = pad(us, 6).replace(/0+$/, "");
    out += `.${frac}`;
  }
  if (isTz) out += "+00";
  out += era;
  return out;
}

// renderTimestamp renders a timestamp value to its canonical text.
export function renderTimestamp(micros: bigint): string {
  return renderDatetime(micros, false);
}

// renderTimestamptz renders a timestamptz value to its canonical text (always UTC, fixed +00).
export function renderTimestamptz(micros: bigint): string {
  return renderDatetime(micros, true);
}
